src/main/java/io/epirus/console/deploy/DeployRunner.java

```diff
@@ -0,0 +1,158 @@
+                                    CliConfig.getConfig(
+                                            CliConfig.getDefaultEpirusConfigPath().toFile()),
+                                    new OkHttpClient()))
+                    .deploy();
+        } else {
+            Console.exitError(USAGE);
+        }
+    }
+
+    public DeployRunner(Network network, AccountManager accountManager, Path workingDirectory) {
+        this.workingDirectory = workingDirectory;
+        this.network = network;
+        Path walletPath = ProjectUtils.loadProjectWalletFile(workingDirectory).get();
+        Path walletPasswordPath =
+                ProjectUtils.loadProjectPasswordWalletFile(workingDirectory).get();
+        this.credentials = ProjectUtils.createCredentials(walletPath, walletPasswordPath);
+        this.accountManager = accountManager;
+    }
+
+    public DeployRunner(Network network, AccountManager accountManager) {
+        this.workingDirectory = Paths.get(System.getProperty("user.dir"));
+        this.network = network;
+        Path walletPath = ProjectUtils.loadProjectWalletFile(workingDirectory).get();
+        Path walletPasswordPath =
+                ProjectUtils.loadProjectPasswordWalletFile(workingDirectory).get();
+        this.credentials = ProjectUtils.createCredentials(walletPath, walletPasswordPath);
+        this.accountManager = accountManager;
+    }
+
+    public void deploy() throws Exception {
+        AccountUtils.accountInit(accountManager);
+        this.accountManager.checkIfAccountIsConfirmed();
+        fundWallet();
+        waitForBalanceUpdate();
+        uploadSolidityMetadata(network, workingDirectory);
+        runGradle(workingDirectory);
+    }
+
+    private void fundWallet() {
+        int tries = 5;
+        while (tries-- > 0) {
+            try {
+                BigInteger accountBalance = accountManager.getAccountBalance(credentials, network);
+                if (accountBalance.equals(BigInteger.ZERO)) {
+                    WalletFunder.fundWallet(
+                            credentials.getAddress(),
+                            Faucet.valueOf(network.getNetworkName().toUpperCase()),
+                            "alex");
+                }
+                return;
+            } catch (Exception e) {
+
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private void waitForBalanceUpdate() {
+        System.out.println("Checking the account balance...");
+        int tries = 5;
+        while (tries-- > 0) {
+            try {
+                BigInteger accountBalance = accountManager.getAccountBalance(credentials, network);
+                if (!accountBalance.equals(BigInteger.ZERO)) {
+                    return;
+                } else {
+                    Thread.sleep(10000);
```

Comment:
The number of retries and the sleep interval should have some sort of relationship with the block time maybe? Instead of sleeping like this, can we not poll and wait for the next block?

So if after 5 blocks the wallet is still not funded, it's fine to give up. Also probably a good idea to have some form of timeout, in cases like when rinkeby is broken and not producing new blocks.

Also, like pointed out for the previous try and sleep loop, we're here sleeping for 10 seconds even if we know we're not going to retry.

Reply:
Ideally yes we would like to check if the account has been funded and we would like to do so over the next five blocks. The issue with this is that there is currently a way that i am aware in web3j to do this. On rinkeby the average block time now is 15s but on ropsten is a lot lower.